### PR TITLE
Call _deploy during deploy and update

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/1932

### CODE REVIEW

File: src/neo/SmartContract/ApplicationEngine.Contract.cs [hunk 1]
```diff
@@ -50,6 +50,34 @@ protected internal ContractState CreateContract(byte[] script, byte[] manifest)
             if (!contract.Manifest.IsValid(hash)) throw new InvalidOperationException($"Invalid Manifest Hash: {hash}");
 
             Snapshot.Contracts.Add(hash, contract);
+
+            // Execute _deploy
+
+            ContractMethodDescriptor md = contract.Manifest.Abi.GetMethod("_deploy");
```
Comment (devhawk) [inline src/neo/SmartContract/ApplicationEngine.Contract.cs:56]: will NEON support a method named _deploy ?

Comment (erikzhang) [reply src/neo/SmartContract/ApplicationEngine.Contract.cs:56]: Yes.

**Comment by erikzhang** [comment]
```c#
void _deploy(bool updated);
```

@devhawk @shargon What do you think? If it is the first deployment, then `updated` is `false`. If it is an update, then `updated` is `true`.

File: src/neo/SmartContract/ApplicationEngine.Contract.cs (hunk 1 above)
Comment (Tommo-L) [reply src/neo/SmartContract/ApplicationEngine.Contract.cs:56]: 1. Do we need to call `_deploy` in `System.Contract.Update` syscall?
2. Rename `System.Contract.Create` to `System.Contract.Deploy` ?

File: src/neo/SmartContract/ApplicationEngine.Contract.cs [hunk 2]
```diff
@@ -50,6 +50,13 @@ protected internal ContractState CreateContract(byte[] script, byte[] manifest)
...
+            if (md != null)
+                CallContractInternal(contract, md, new Array(ReferenceCounter), CallFlags.All);
+
             return contract;
```
Comment (erikzhang) [inline src/neo/SmartContract/ApplicationEngine.Contract.cs:60]: The return value is pushed onto the current stack, which is the stack of `_deploy()`. I think we should push it onto the caller's stack.

**Comment by devhawk** [comment]
I'd like to know more about what the NEON developer experience will be for authoring the _deploy method. Just write a method named _deploy? Attribute? Are we going to check to ensure the method has the correct signature?

Otherwise, looks good.

File: src/neo/SmartContract/ApplicationEngine.Contract.cs (hunk 2 above)
Comment (shargon) [reply src/neo/SmartContract/ApplicationEngine.Contract.cs]: Done

**Comment by erikzhang** [comment]
https://github.com/neo-project/neo/blob/64c3a870cdf5026a506c85f4eb3335608cd6e0c5/src/neo/SmartContract/ApplicationEngine.cs#L100-L104

`_deploy()` pushes a `null` onto the stack, which should be removed.

File: src/neo/SmartContract/ApplicationEngine.Contract.cs [hunk 3]
```diff
@@ -50,7 +50,18 @@ protected internal ContractState CreateContract(byte[] script, byte[] manifest)
...
+
+            // We should push it onto the caller's stack.
+
+            Push(Convert(contract));
+
+            // Execute _deploy
+
+            ContractMethodDescriptor md = contract.Manifest.Abi.GetMethod("_deploy");
+            if (md != null)
+                CallContractInternal(contract, md, new Array(ReferenceCounter) { false }, CallFlags.All, false);
```
Comment (erikzhang) [inline src/neo/SmartContract/ApplicationEngine.Contract.cs:62]: If you set `NeedCheckReturnValue` to `false`, the method (`_deploy`) can push any values onto the stack.

Comment (shargon) [reply src/neo/SmartContract/ApplicationEngine.Contract.cs:62]: How do you think that we can remove the null value?

Comment (shargon) [reply src/neo/SmartContract/ApplicationEngine.Contract.cs]: @erikzhang please take a look again, it could solve the problem

**Review by erikzhang — APPROVED** [review]

**Pull request merged**